convert2rhel/subscription.py

```diff
@@ -277,28 +287,103 @@ def args(self):
 
         return args
 
-    def __call__(self):
+    @property
+    def connection_opts(self):
         """
-        Run the subscription-manager command.
+        This property is a dbus.Dictionary that contains the connection options for RHSM
+        dbus calls.
 
-        Wrapper for running the subscription-manager command that keeps
-        secrets secure.
+        Set :attr:`server_url` to affect this value.
         """
-        if self.password:
-            loggerinst.debug(
-                "Calling command '%s %s'" % (self.cmd, " ".join(hide_secrets(self.args)))
-            )  # lgtm[py/clear-text-logging-sensitive-data]
-            output, ret_code = utils.run_cmd_in_pty(
-                [self.cmd] + self.args, expect_script=(("[Pp]assword: ", self.password + "\n"),), print_cmd=False
-            )
-        else:
-            # Warning: Currently activation_key can only be specified on the CLI. This is insecure
-            # but there's nothing we can do about it for now. Once subscription-manager issue:
-            # https://issues.redhat.com/browse/ENT-4724 is implemented, we can change both password
-            # and activation_key to use a file-based approach to passing the secrets.
-            output, ret_code = utils.run_subprocess([self.cmd] + self.args, print_cmd=False)
+        connection_opts = {}
 
-        return output, ret_code
+        if self.server_url:
+            url_parts = urllib.parse.urlsplit(self.server_url, allow_fragments=False)
+            connection_opts["host"] = url_parts.hostname
+            if url_parts.port is not None:
+                connection_opts["port"] = url_parts.port
+
+            # Current subscription-manager server-side does not take query and fragments at this
+            # path but include them for future proofing.
+            path = urllib.parse.urlunsplit(("", "", url_parts.path, url_parts.query, url_parts.fragment))
+            connection_opts["handler"] = path
```

Comment:
I'd simply leave them out; it would require work in sub-man to support them, if ever, so let's keep things simple here and use only `url_parts.path`

Reply:
<nod> I can see your point of view as a subscription-manager dev.  From my point of view, though, I'd rather code something now that might work if subscription-manager should choose to start using query and fragment in the future while the way that subscription-manager and dbus work is fresh in my mind.  If subscription-manager never needs to add query or fragment here, then there is no harm.  If subscription-manager adds them in the way that the code anticipates, then I won't have to try to figure out the mapping of cli->dbus args again.  If subscription-manager adds them in a way that the code doesn't anticipate, at least when someone looks at the code they'll see all the pieces even if they're assembled in the wrong way.

Reply:
I don't think subscription-manager is going to get that functionality any time soon. It is not only about taking query and fragment from an URL passed to `--serverurl`, but also needs new configuration bits (as hostname, port, etc are stored separately there) and all the associated machinery for it, and there is simply nothing that ATM needs them.

So, while I understand your POV, I'd keep things simple and not adding things that could misbehave.